compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/GBKStreamingTransform.java

```diff
@@ -0,0 +1,337 @@
+        checkAndInvokeBundle();
+        final KV<K, InputT> kv = element.getValue();
+        final KeyedWorkItem<K, InputT> keyedWorkItem =
+          KeyedWorkItems.elementsWorkItem(kv.getKey(),
+            Collections.singletonList(element.withValue(kv.getValue())));
+        getDoFnRunner().processElement(WindowedValue.valueInGlobalWindow(keyedWorkItem));
+        checkAndFinishBundle();
+      } catch (final Exception e) {
+        e.printStackTrace();
+        throw new RuntimeException("exception trigggered element " + element.toString());
+      }
+  }
+
+  /**
+   * Process the collected data, trigger timers, and emit watermark to downstream operators.
+   * @param processingTime processing time
+   * @param synchronizedTime synchronized time
+   * @param triggerWatermark watermark
+   */
+  private void processElementsAndTriggerTimers(final Instant processingTime,
+                                               final Instant synchronizedTime,
+                                               final Watermark triggerWatermark) {
+    triggerTimers(processingTime, synchronizedTime, triggerWatermark);
+    emitOutputWatermark();
+  }
+
+  /**
+   * Emit watermark to downstream operators.
+   * Output watermark = max(prev output watermark, min(input watermark, watermark holds)).
+   */
+  private void emitOutputWatermark() {
+    // Find min watermark hold
+    Watermark minWatermarkHold = keyAndWatermarkHoldMap.isEmpty()
+      ? new Watermark(dataReceived ? Long.MIN_VALUE : Long.MAX_VALUE)
+      : Collections.min(keyAndWatermarkHoldMap.values());
+
+    Watermark outputWatermarkCandidate = new Watermark(
+      Math.max(prevOutputWatermark.getTimestamp(),
+        Math.min(minWatermarkHold.getTimestamp(), inputWatermark.getTimestamp())));
+
+    while (outputWatermarkCandidate.getTimestamp() > prevOutputWatermark.getTimestamp()) {
+      // Progress
+      prevOutputWatermark = outputWatermarkCandidate;
+      // Emit watermark
+      getOutputCollector().emitWatermark(outputWatermarkCandidate);
+      // Remove minimum watermark holds
+      if (minWatermarkHold.getTimestamp() == outputWatermarkCandidate.getTimestamp()) {
+        final long minWatermarkTimestamp = minWatermarkHold.getTimestamp();
+        keyAndWatermarkHoldMap.entrySet()
+          .removeIf(entry -> entry.getValue().getTimestamp() == minWatermarkTimestamp);
+      }
+
+      minWatermarkHold = keyAndWatermarkHoldMap.isEmpty()
+        ? new Watermark(Long.MAX_VALUE) : Collections.min(keyAndWatermarkHoldMap.values());
+
+      outputWatermarkCandidate = new Watermark(
+        Math.max(prevOutputWatermark.getTimestamp(),
+          Math.min(minWatermarkHold.getTimestamp(), inputWatermark.getTimestamp())));
+    }
+  }
+
+  /**
+   * Trigger timers that need to be fired.
+   * @param watermark watermark
+   */
+  @Override
+  public void onWatermark(final Watermark watermark) {
+    if (watermark.getTimestamp() <= inputWatermark.getTimestamp()) {
+      return;
```

Comment:
How about throwing a RuntimeException? In fact, this should not happen

Reply:
I agree. Fixed it.